### PR TITLE
Alerting: Stop firing of alert when it is updated

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
Currently, when a user creates an alert (Grafana hosted), the alert starts firing, and then the user modifies it by changing labels, the user can see two firing alerts at the same time: 
- the previous alert is still firing because its end time is set to 2 evaluation intervals in the future (but not less than 30 seconds)
- depending on the configuration, the new version of the alert is promoted from pending to firing and starts firing alongside with the old one. 

Same situation happens when user deletes an alert rule that has firing instances. The alertmanager keeps firing the alert until it expires.

Current implementation, cleans up alert state but does not expire alerts in Alertmanager: when user submits update or delete request, API just cleans up the state manager.

This PR changes the behavior. When alert rule is changed or deleted, all firing instances are immediately expired along with state. This is done by moving the responsibility to clean up state and update Alertmanager to scheduler. 

- Scheduler is updated to expose two new public functions `UpdateAlertRule` and `DeleteAlertRule`. 
- Rule evaluation routine is updated to read a new channel `updateCh` which is used by the functions to let know the routine loop to update the rule. 
- rule evaluation routine is updated to reset state and send expired alert instances to Alertmanager every time version of alert changes.

**Which issue(s) this PR fixes**:
Related https://github.com/grafana/grafana/issues/39154

**Special notes for your reviewer**:
- The PR does not fix the issue above. It addresses only clearing up alerts. There is one thing that is missing: when a user updates an alert rule, API should check whether there are any changes and do not do anything if the rule is not changed. This will be addressed in another PR.

### CODE REVIEW

**Review by grobinson-grafana — COMMENTED** [review]
I think this is OK, but I want to get a review from @gotjosh too!

**Comment by yuri-tceretian** [comment]
There is one major issue with the implementation. It does not handle external alertmanager. Switching back to the draft.

**Comment by gotjosh** [comment]
@yuri-tceretian our test coverage in the scheduler are very poor at the moment, and the test themselves don't seem to have much to say about the scheduling of alerts themselves (intervals/holds, when to fire or not, etc).

How do you feel about adding coverage to the existing logic, before we delve into modifications of this size? I think that would give us more confidence and will pay off down the line.

**Comment by yuri-tceretian** [comment]
> How do you feel about adding coverage to the existing logic, before we delve into modifications of this size? I think that would give us more confidence and will pay off down the lin

Yes, I will add some tests. Just wanted you to take a look to make sure that the direction is correct.

**Comment by yuri-tceretian** [comment]
I extracted refactoring bit (first few commits)  from this PR to https://github.com/grafana/grafana/pull/40238. Will rebase to main after it is merged.

File: pkg/services/ngalert/schedule/testing.go
Comment (alexweav) [inline pkg/services/ngalert/schedule/testing.go:82]: (question for my own enlightment)

What is the rationale behind using handwritten fakes instead of mocks? It feels like we wouldn't have to manually maintain logic like this if we used a mock which was set up on a per-test or per-test-file basis.

Although some Go mocks require code generators, it looks like there is already precedent for generating such mocks in other packages. For instance, I see generator commands which call https://github.com/golang/mock. Is there some other historical context behind this choice?

Comment (alexweav) [reply pkg/services/ngalert/schedule/testing.go:82]: To be clear, I don't disagree with the use of fakes. Especially since this PR mains consistency with the existing test infrastructure. This was more to fill in context.

File: pkg/services/ngalert/schedule/schedule_unit_test.go
Comment (alexweav) [inline pkg/services/ngalert/schedule/schedule_unit_test.go:904]:
```suggestion
func createFakeScheduler(t *testing.T) *schedule {
```

_Incredibly_ nitpicky comment. These tests look great overall.

File: pkg/services/ngalert/schedule/schedule.go
Comment (alexweav) [inline pkg/services/ngalert/schedule/schedule.go:327]:
```suggestion
		sch.log.Info("unable to delete alert rule routine information by key", "uid", key.UID, "org_id", key.OrgID)
```

Also very nitpicky, but in the previous wording I suspect someone debugging this might confuse this log message for being associated with a read rather than a delete.

Also, the old log specifically assumed the only way the delete could fail would be that the record didn't exist. If this is true, it may be worth adjusting the statement to mention this, e.g. `Unable to delete rule routine information because it did not exist.`.At first glance I wondered why this log wasn't at error level based on the message, but it was more clear after reading the original context.

File: pkg/services/ngalert/schedule/compat.go
Comment (alexweav) [inline pkg/services/ngalert/schedule/compat.go:130]: ...or eval.Error or eval.NoData

Is there any reason we couldn't stop the alert regardless of its state? It seems like in the event of an update we'd always want to get rid resources attached to the old alert. Having this behavior be orthogonal and always happen is a bit easier to reason about.

(Unless "Stopped" has a special meaning in the DB and this would cause the non-firing alert would show up in an unexpected place. If this is the case, this is probably fine. It'd be nice if other components were smart enough to avoid such an invalid transition if it were to exist, though.)

File: pkg/services/ngalert/schedule/testing.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/testing.go:82]: I agree that we need to use mocks but the reality is that we do not do that for now. I am not aware of any decisions we made in past about mocks vs fake. Mocks just are not really popular among developers.

File: pkg/services/ngalert/schedule/compat.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/compat.go:130]: This package decides whether `State` should be converted to `PostableAlert`. Other functions of this package ignore state Normal and Pending as it does not need an alert to be created in the Alertmanager. The scheduler that gets states from the state manager should not be aware of such peculiarities. 
I forgot to update the comment when I updated this function. Previously, when this PR was created, we handled only Alerting state. After https://github.com/grafana/grafana/pull/40540 and https://github.com/grafana/grafana/pull/41869 we start treating NoData and Error as alerting states.

File: pkg/services/ngalert/schedule/schedule_unit_test.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule_unit_test.go:904]: fake is not good either because we do not create a fake scheduler, we create a scheduler that is set up with fake stores. Anyway, I think I've found a better name for it.

File: pkg/services/ngalert/schedule/testing.go
Comment (gerobinson) [reply pkg/services/ngalert/schedule/testing.go:82]: Just a personal opinion here as I have used both `mockgen` and hand written fakes in the past and much prefer the latter for simple mocking. Mockgen does make more sense for more complex mocks that need expectations.

Mockgen can be slow if there are a lot of mocks to generate and it can be harder to debug errors caused when the mock is out of date (i.e. consider changing an interface and forgetting to regenerate the mocks). That said, for more complex mocks where we want expectations it makes sense to use them. I just don't think we should use them all the time for all tests. :)

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:567]: I am thinking if we can make it more apparent that both `<-updateCh` and `<-evalCh` have the same prelude (i.e. both of them call `updateRule`, write the same log lines with the same message, and update `currentRule`. I think we could even remove `updateRule` from `<-evalCh` if we check `<-updateCh` first?

Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:567]: I thought about that too but then could find a use case when this is not true - Grafana HA mode.

File: pkg/services/ngalert/schedule/compat.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/compat.go:132]: How come we use `clock.Clock` instead of `time.Time`?

File: pkg/services/ngalert/schedule/schedule.go
Comment (gerobinson) [inline pkg/services/ngalert/schedule/schedule.go:434]: Do we have a test to show that when we delete an alert rule with `DeleteAlertRule` the scheduler will still call `sch.DeleteAlertRule`? I ask because if we call `DeleteAlertRule` then won't we remove the alert rule from `sch.registry` before the scheduler has called `sch.registry.KeyMap()` such that we never call `sch.DeleteAlertRule(key)`?

File: pkg/services/ngalert/schedule/compat.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/compat.go:132]: I use clock because we have tooling to mock it. Same as in scheduler.

File: pkg/services/ngalert/schedule/schedule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:434]: Sorry, I do not understand the use-case you describe. Can you please clarify. As for tests, there are few tests https://github.com/grafana/grafana/pull/39975/files/6af0053ca4c234154c9629c233d3246f48be0bb0#diff-bd46cfcb4ef816c767bb8b9d3637680e1c6ac38f7fd66f4d7635b06adc6d588fR866

**Review by grobinson-grafana — APPROVED** [review]

**Pull request merged**
